### PR TITLE
fix(glusterfs): Using fixed length for endpointname and labels since they must be no more than 63 characters

### LINKED ISSUE #56379: glusterfs endpoint/svc name and labels should no more than 63 characters
**Is this a BUG REPORT or FEATURE REQUEST?**: BUG REPORT

> Uncomment only one, leave it on its own line: 
>
> /kind bug
> /kind feature

/kind bug

**What happened**:

If I use glusterfs provisioner with a pvc name more than 63 characters, provision will fail

the pvc like this:

```
[root@c2v98 ~]# kubectl get pvc -o yaml pvc-123456789012345678901234567890123456789012345678901234567890
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    volume.beta.kubernetes.io/storage-class: heketi-storageclass
    volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/glusterfs
  creationTimestamp: 2017-11-26T16:39:33Z
  name: pvc-123456789012345678901234567890123456789012345678901234567890
  namespace: default
  resourceVersion: "302171"
  selfLink: /api/v1/namespaces/default/persistentvolumeclaims/pvc-123456789012345678901234567890123456789012345678901234567890
  uid: 61e23f5e-d2c8-11e7-94d9-525400461b75
spec:
  accessModes:
  - ReadWriteMany
  resources:
    requests:
      storage: 1Gi
status:
  phase: Pending
```

and got those events:

```
[root@c2v98 ~]# kubectl describe pvc pvc-123456789012345678901234567890123456789012345678901234567890
Name:          pvc-123456789012345678901234567890123456789012345678901234567890
Namespace:     default
StorageClass:  heketi-storageclass
Status:        Pending
Volume:        
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-class=heketi-storageclass
               volume.beta.kubernetes.io/storage-provisioner=kubernetes.io/glusterfs
Capacity:      
Access Modes:  
Events:
  Type     Reason              Age               From                         Message
  ----     ------              ----              ----                         -------
  Warning  ProvisioningFailed  21s (x2 over 1m)  persistentvolume-controller  Failed to provision volume with StorageClass "heketi-storageclass": create volume error: failed to create endpoint/service error creating endpoint: Endpoints "glusterfs-dynamic-pvc-123456789012345678901234567890123456789012345678901234567890" is invalid: metadata.labels: Invalid value: "pvc-123456789012345678901234567890123456789012345678901234567890": must be no more than 63 characters
```

**What you expected to happen**:

The pv should be created if pvc name length is allowed (even more than 63 characters)

**How to reproduce it (as minimally and precisely as possible)**:

Create a pvc with name more than 63 characters

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.8.4
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`): 
- Install tools: kubeadm
- Others:

### PR DESCRIPTION
**Pull request by ijumps**
**What this PR does / why we need it**:  Using fixed length for endpointname and labels since they must be no more than 63 characters

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #56379

**Special notes for your reviewer**:

Since we need an unique name for glusterfs  ep/svc, and the length must be no more than 63 characters, a quick fix is using the pvc UUID, which has a fixed length and always be unique.

As for upgrade, the endpoints was stored in glusterfs pv spec, changing glusterfs naming has no affect on existing glusterfs pv.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/volume/glusterfs/glusterfs.go [hunk 1]
```diff
@@ -763,13 +763,13 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsVolum
 		return nil, 0, fmt.Errorf("error [%v] when getting cluster nodes for volume %s", err, volume)
 	}
 
-	// The 'endpointname' is created in form of 'glusterfs-dynamic-<claimname>'.
+	// The 'endpointname' is created in form of 'glusterfs-dynamic-<claimuid>'.
 	// createEndpointService() checks for this 'endpoint' existence in PVC's namespace and
 	// If not found, it create an endpoint and svc using the IPs we dynamically picked at time
 	// of volume creation.
-	epServiceName := dynamicEpSvcPrefix + p.options.PVC.Name
+	epServiceName := dynamicEpSvcPrefix + string(p.options.PVC.UID)
```
Comment (dixudx) [inline pkg/volume/glusterfs/glusterfs.go]: I don't think this will help solve the issue.

/cc @msau42 @saad-ali

Comment (ijumps) [reply pkg/volume/glusterfs/glusterfs.go]: @dixudx I tested, and this works. I add some explain in Special notes

File: pkg/volume/glusterfs/glusterfs.go [hunk 2]
```diff
@@ -797,7 +797,7 @@ func (p *glusterfsVolumeProvisioner) createEndpointService(namespace string, epS
 			Namespace: namespace,
 			Name:      epServiceName,
 			Labels: map[string]string{
-				"gluster.kubernetes.io/provisioned-for-pvc": pvcname,
+				"gluster.kubernetes.io/provisioned-for-pvc": pvcID,
```
Comment (jsafrane) [inline pkg/volume/glusterfs/glusterfs.go]: Please use different label for UID, `gluster.kubernetes.io/provisioned-for-pvc-uid`?
And add the PVC name to annotation, so user can see what created this endpoint/service.

File: pkg/volume/glusterfs/glusterfs.go (hunk 1 above)
Comment (saad-ali) [reply pkg/volume/glusterfs/glusterfs.go]: Will this break backwards compatibility?

Comment (msau42) [reply pkg/volume/glusterfs/glusterfs.go]: It shouldn't because the name is stored in the PV spec, and the deletion uses the name from the PV.

Comment (dixudx) [reply pkg/volume/glusterfs/glusterfs.go]: Agree @msau42

**Comment by ijumps** [comment]
> Please use different label for UID, gluster.kubernetes.io/provisioned-for-pvc-uid?
And add the PVC name to annotation, so user can see what created this endpoint/service.

@jsafrane Good idea to keep this in annotation. Updated.

> This is semi-unrelated, but how come in Delete(), only the service is deleted, and not the endpoints object?

@msau42 AFAIK, `ep` will be deleted when `svc` is deleted.

**Comment by humblec** [comment]
@ijumps @msau42  One thing to note here is, many admins has scripts to fetch the subjected service/endpoint name based on the pvc name. Those are not going to work after this change. If its a concern on fixed length for endpointname and labels why we cant add a check on existing naming schema ?

**Comment by jsafrane** [comment]
> why we cant add a check on existing naming schema 

What would the code do if the PVC name was too long? Use UID? Then we have half of services with UIDs and second half with PVC names. That's IMO even worse than using UID consistently.

**Comment by ijumps** [comment]
> If its a concern on fixed length for endpointname and labels why we cant add a check on existing naming schema ?

@humblec If you mean set limitation on pvc name length, I don't think this's an enough reason to break backwards compatibility.

> What would the code do if the PVC name was too long?

I agree with @jsafrane to use UID consistently

@jsafrane @msau42 Any suggestions?

**Comment by humblec** [comment]
```     
What would the code do if the PVC name was too long? Use UID? Then we have half of services with UIDs and second half with PVC names. That's IMO even worse than using UID consistently.
```

@humblec If you mean set limitation on pvc name length, I don't think this's an enough reason to break backwards compatibility.
```
@jsafrane @msau42 what I meant was neither  using `UID` nor set limitation on PVC name length, rather cut the pvc name length with a limit allowed on endpoint name from plugin code. ie, if PVC name is `abc............z`, use till `abc..p` and discard `p...z`. atleast this will help `grep` scripts to an extent.Thoughts ?

**Comment by jsafrane** [comment]
> cut the pvc name length with a limit allowed on endpoint name from plugin code.

What if two PVC names are the same except the last character that you cut here?

**Pull request closed without merging**